daemon/logger/local/read.go

```diff
@@ -99,53 +98,65 @@ func getTailReader(ctx context.Context, r loggerutils.SizeReaderAt, req int) (io
 type decoder struct {
 	rdr   io.Reader
 	proto *logdriver.LogEntry
-	buf   []byte
+	// buf keeps bytes from rdr.
+	buf []byte
+	// offset is the position in buf.
+	// If offset > 0, buf[offset:] has bytes which are read but haven't used.
+	offset int
+	// msgLen is the length of the next log message.
+	// If msgLen = 0, a new value must be read from rdr.
+	msgLen int
 }
 
-func (d *decoder) Decode() (*logger.Message, error) {
-	if d.proto == nil {
-		d.proto = &logdriver.LogEntry{}
-	} else {
-		resetProto(d.proto)
-	}
-	if d.buf == nil {
-		d.buf = make([]byte, initialBufSize)
-	}
-	var (
-		read int
-		err  error
-	)
-
+func (d *decoder) readRecord(size int) error {
+	var err error
 	for i := 0; i < maxDecodeRetry; i++ {
 		var n int
-		n, err = io.ReadFull(d.rdr, d.buf[read:encodeBinaryLen])
+		n, err = io.ReadFull(d.rdr, d.buf[d.offset:size])
+		d.offset += n
 		if err != nil {
 			if err != io.ErrUnexpectedEOF {
-				return nil, errors.Wrap(err, "error reading log message length")
+				return err
 			}
-			read += n
 			continue
 		}
-		read += n
 		break
 	}
 	if err != nil {
-		return nil, errors.Wrapf(err, "could not read log message length: read: %d, expected: %d", read, encodeBinaryLen)
+		return err
 	}
+	d.offset = 0
+	return nil
+}
 
-	msgLen := int(binary.BigEndian.Uint32(d.buf[:read]))
-
-	if len(d.buf) < msgLen+encodeBinaryLen {
-		d.buf = make([]byte, msgLen+encodeBinaryLen)
+func (d *decoder) Decode() (*logger.Message, error) {
+	if d.proto == nil {
+		d.proto = &logdriver.LogEntry{}
 	} else {
-		if msgLen <= initialBufSize {
-			d.buf = d.buf[:initialBufSize]
+		resetProto(d.proto)
+	}
+	if d.buf == nil {
+		d.buf = make([]byte, initialBufSize)
+	}
+
+	if d.msgLen == 0 {
+		err := d.readRecord(encodeBinaryLen)
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not read a size header")
+		}
+		msgLen := int(binary.BigEndian.Uint32(d.buf[:encodeBinaryLen]))
+		if len(d.buf) < msgLen+encodeBinaryLen {
+			d.buf = make([]byte, msgLen+encodeBinaryLen)
 		} else {
-			d.buf = d.buf[:msgLen+encodeBinaryLen]
+			if msgLen <= initialBufSize {
+				d.buf = d.buf[:initialBufSize]
+			} else {
+				d.buf = d.buf[:msgLen+encodeBinaryLen]
+			}
 		}
```

Comment:
This buffer resizing logic could move up into `readRecord`.  We could then additionally make it so that code reading `d.buf` explicitly tries to read `d.buf[:size]` rather than the entire buffer, meaning we don't have to reslice `d.buf` making it smaller for small records and then having to re-grow it later (e.g. we can conserve the single buffer and reuse it for the life of the `decoder`, only growing it once necessary).  (The downside of this is that a single large message means we grow the buffer up to that maximum size and may not ever need it that big again and we're preventing Go from garbage-collecting that unused portion.)

@cpuguy83 What do you think about that?

Reply:
Does the log record have its upper limit in terms of the size? Bit worried about the downside you've mentioned.